src/VisualStudio/Core/Def/DocumentOutline/DocumentOutlineHelper.cs

```diff
@@ -0,0 +1,366 @@
+            SortOption sortOption,
+            ITextSnapshot? oldSnapshot,
+            ITextSnapshot? currentSnapshot)
+        {
+            // We want to log which sort option is used
+            Logger.Log(sortOption switch
+            {
+                SortOption.Name => FunctionId.DocumentOutline_SortByName,
+                SortOption.Order => FunctionId.DocumentOutline_SortByOrder,
+                SortOption.Type => FunctionId.DocumentOutline_SortByType,
+                _ => throw new NotImplementedException(),
+            });
+
+            // Sort the top-level DocumentSymbolViewModels
+            var sortedDocumentSymbolModels = sortOption switch
+            {
+                SortOption.Name => documentSymbolModels.Sort((x, y) => x.Name.CompareTo(y.Name)),
+                SortOption.Order => documentSymbolModels.Sort((x, y) =>
+                {
+                    if (oldSnapshot is null)
+                        throw new ArgumentNullException(nameof(oldSnapshot));
+                    else if (currentSnapshot is null)
+                        throw new ArgumentNullException(nameof(currentSnapshot));
+
+                    return CompareSymbolOrder(x, y, oldSnapshot, currentSnapshot);
+                }),
+                SortOption.Type => documentSymbolModels.Sort(CompareSymbolType),
+                _ => throw new NotImplementedException()
+            };
+
+            // Recursively sort descendant DocumentSymbolViewModels
+            foreach (var documentSymbolModel in sortedDocumentSymbolModels)
+            {
+                documentSymbolModel.Children = Sort(documentSymbolModel.Children, sortOption, oldSnapshot, currentSnapshot);
+            }
+
+            return sortedDocumentSymbolModels;
+        }
+
+        /// <summary>
+        /// Returns an immutable array of DocumentSymbolViewModels such that each node or one of its descendants matches the given pattern.
+        /// </summary>
+        public static ImmutableArray<DocumentSymbolViewModel> Search(ImmutableArray<DocumentSymbolViewModel> documentSymbolModels, string pattern)
+        {
+            var documentSymbols = ArrayBuilder<DocumentSymbolViewModel>.GetInstance();
+            var patternMatcher = PatternMatcher.CreatePatternMatcher(pattern, includeMatchedSpans: false, allowFuzzyMatching: true);
+
+            foreach (var documentSymbol in documentSymbolModels)
+            {
+                if (SearchNodeTree(documentSymbol, patternMatcher))
+                    documentSymbols.Add(documentSymbol);
+            }
+
+            return documentSymbols.ToImmutableAndFree();
+        }
+
+        /// <summary>
+        /// Returns true if the name of one of the tree nodes results in a pattern match.
+        /// </summary>
+        public static bool SearchNodeTree(DocumentSymbolViewModel tree, PatternMatcher patternMatcher)
+        {
+            var pattern = patternMatcher.Matches(tree.Name);
```

Comment:
we never use this variable. This causes the `Correctness_Build` part of CI to fail with

```
error IDE0059: Unnecessary assignment of a value to 'pattern'
```

I believe we should just remove this?